### PR TITLE
Sets

### PR DESCRIPTION
**Pull request by NelliMg**

### CODE REVIEW

File: src/sets/model/MyCoffee.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+/**
+ * Class Coffee
+ * Which has two fields
+ * 1)Coffee type coffee
+ * 2)Integer type price
+ * functions equals hashCode and compareTo were overriden
+ */
+public class MyCoffee implements Comparable<MyCoffee>{
+    private Coffee coffee;
```
Comment (DiannaAsatryan) [inline src/sets/model/MyCoffee.java]: it is better to rename it to CoffeeType as Coffee is confusing

File: src/sets/implementation/SearchTree.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+        return current;
+    }
+
+    /**
+     * recieves the item to be added to the tree
+     * @param item
+     * @return false if the item existed/true if it didn't exist and was successfully added
+     */
+    public boolean add(@NotNull T item) {
+        if(search(item)){
```
Comment (DiannaAsatryan) [inline src/sets/implementation/SearchTree.java]: you are doing duplicated work here, don't need to search element, as your add also is traversing over three to find an appropriate place to add

File: src/sets/implementation/MySet.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+package sets.implementation;
+
+import org.jetbrains.annotations.NotNull;
+
+public class MySet<T extends Comparable<T>> {
```
Comment (DiannaAsatryan) [inline src/sets/implementation/MySet.java]: better is to give Java DOCs for your API which client must use(in this case MySet)

File: src/sets/implementation/MySet.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    public MySet() {
+        this.setTree = new SearchTree<>();
+    }
+    public boolean add(@NotNull T item){
+        return setTree.add(item);
+    }
+    public boolean contains(@NotNull T item){
+        return setTree.search(item);
+    }
+    public boolean remove(@NotNull T item){
```
Comment (DiannaAsatryan) [inline src/sets/implementation/MySet.java]: do not use annotations which requires  additional configuration on my side like @NotNull

File: src/sets/implementation/SearchTree.java [hunk 5]
```diff
@@ -0,0 +1,203 @@
...
+     * @return the new node if current was null and the same Node if it wasn't null
+     */
+    private Node recursiveAdd(Node current, T item) {
+        if (current == null) {
+            return new Node(item);
+        }else if (current.data.compareTo(item) > 0) {
+            current.left = recursiveAdd(current.left, item);
+        } else if (current.data.compareTo(item) < 0) {
+            current.right = recursiveAdd(current.right, item);
+        }
```
Comment (DiannaAsatryan) [inline src/sets/implementation/SearchTree.java:53]: I guess you can just add here one else case with exception and cover (exist) scenario

File: src/sets/implementation/SearchTree.java [hunk 6]
```diff
@@ -0,0 +1,203 @@
...
+
+import java.util.Iterator;
+
+/**
+ * Generic search tree implementation
+ * that if Traversed from left to right will bew sorted in ascending array
+ * @param <T>
+ */
+
+public class SearchTree<T extends Comparable<T>> implements Iterable<T> {
```
Comment (DiannaAsatryan) [inline src/sets/implementation/SearchTree.java:13]: I think your MySet class should implement Iterable as we need to traverse on it

**Review by DiannaAsatryan — APPROVED** [review]
Well done Nelly jan, I think you have implemented yurself which is appreciated. Just small changes to consider

**Comment by NelliMg** [comment]
> Well done Nelly jan, I think you have implemented yurself which is appreciated. Just small changes to consider



> Well done Nelly jan, I think you have implemented yurself which is appreciated. Just small changes to consider

Thank you:) I made the changes and got rid of duplicate operations

**Pull request merged**